### ISSUE TITLE
How to Escape/Sanitise Untrusted Data/Params in SQL Queries to avoid SQL Injection in node.js?

### ISSUE THREAD

**Issue by nelsonic** [dwyl/learn-postgresql#64]
# Why?

It would be (_highly_) _irresponsible_ of us to introduce people to SQL+PostgreSQL without covering **SQL _injections_** (SQLi) and how to _defend/protect_ against them. According to the [OWASP Top 10](https://github.com/dwyl/learn-security/issues/37), "**injection**" in _general_ and SQL injection _specifically_ are still the number _one_ source of security vulnerabilities in web apps. 

![image](https://user-images.githubusercontent.com/194400/56095677-213aec00-5ed7-11e9-94d5-bd37c3dce4a0.png)

![image](https://user-images.githubusercontent.com/194400/56095704-5a735c00-5ed7-11e9-8f7c-2afdb55f96c9.png)

> "_Injection flaws, such as SQL, NoSQL, OS, and LDAP injection, occur when **untrusted data** is sent to an interpreter as part of a command or query. The attacker’s hostile data can trick the
interpreter into executing unintended commands or accessing data without proper authorization._"

An SQL injection can be _devastating_ to a web application because it can result in anything from privilege escalation (_where the attacker makes themself an "admin" of your app_) to data theft and database or table _deletion_! 

[![XKCD Bobby Drop Tables](https://user-images.githubusercontent.com/194400/56096026-a83d9380-5eda-11e9-8c6c-70a60d384cc1.png)](https://xkcd.com/327)
https://xkcd.com/327 - _we can laugh at this comic because we haven't suffered an SQLi attack._ 😥 

# What?

When inserting data (_or running a parameterised query_) in Postgres, we must _sanitise_ (_or "escape"_) the parameters even when we _trust_ them!

(_thankfully, **`Ecto`** sanitises all queries automatically+transparently, so when using [**`Phoenix`**](https://github.com/dwyl/learn-phoenix-framework) this is taken care of and we can all just get on with our building our App ... but in Node.js there's more work_)

# How?

Historically we have used the [`pg-escape`](https://github.com/segmentio/pg-escape) library to escape our queries. It worked (_mostly_) and we did not give it any more thought. But while building the example #51 we noted that "_escape.literal throws `val.indexOf is not a function` for **`numbers`** and **`booleans`**_" see: https://github.com/segmentio/pg-escape/issues/15
Which means that we _cannot_ insert INT/BOOL into a Postgres db or select based on the an INT/BOOL parameter ... we've been using the following "work around":

In our case the `person.id` is an `INT` so we write:
```js
const fields = "username, website";
let q = escape(`SELECT %L FROM people WHERE id = $id ORDER BY id ASC LIMIT 1`, fields); 
q = q.replace('$id', parseInt(id, 10)));
```
You'll notice that this does not actually do any escaping of the `id` parameter.
The `$id` variable just gets replaced in the query without any checking/sanitisation.
This is _obviously_ an undesirable situation, so it got me thinking: "_How can we **fix** this?_"

The basic example above is just to illustrate the problem. The reality is _much_ worse! e.g:
```js
function insert_person (data, callback) {
  connect( function insert_person_after_connected () {
    const { name, username, bio, worksfor, location, website, uid,
      stars, followers, following, contribs } = data;
    const fields = `(name, username, bio, worksfor, location, website, uid,
      stars, followers, following, contribs, recent_activity)`;
    let q = escape(`INSERT INTO people %s
      VALUES (%L, %L, %L, %L, %L, %L, $u, $s, $f, $g, $c, $r)`,
      fields, name, username, bio, worksfor, location, website);
      q = q.replace('$u', parseInt(uid, 10))
        .replace('$s', parseInt(stars, 10))
        .replace('$f', parseInt(followers, 10))
        .replace('$g', parseInt(following, 10))
        .replace('$c', parseInt(contribs, 10))
        .replace('$r', utils.recent_activity(data));

    PG_CLIENT.query(q, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}
```
In the case of our `insert_person` function there are ***six*** INT values that have to be "replaced"... 🤦‍♂️ 

As you can imagine, I'm _really_ unhappy with this. (_hence opening this issue to fix it_!)

A bit of searching for "_nodejs postgresql sanitize_" landed me on the SO question:
https://stackoverflow.com/questions/41455585/does-pg-node-postgres-automatically-sanitize-data 

![image](https://user-images.githubusercontent.com/194400/56097290-205f8580-5eea-11e9-95e5-18e6992e493a.png)

The _accepted_ answer in his SO question/answer: "_Does pg (node-postgres) automatically sanitize data_?" is that **`node-postgres`** (_recently renamed **`pg`**_) "_offers **no protection** against SQL injection_" ...

_Thankfully_, @tuananh took the time to write a _factually correct_ follow-up answer informing that the `pg` module does in fact offer SQL injection protection when using the second argument of the `query` method to supply an array of parameters. 

https://github.com/brianc/node-postgres/wiki/FAQ#8-does-node-postgres-handle-sql-injection
![image](https://user-images.githubusercontent.com/194400/56097317-5bfa4f80-5eea-11e9-851b-31ad0ba6778d.png)

> **Note**: the fact that the **_accepted_ answer** on an SO question is **patently _wrong_** is symptomatic of SO's inability to re-assign the **_accepted_ answer** in light of new evidence. 🙄 

Anyway, long story short, we are able to re-write the _hideous_ `insert_person` function 
which previously had _6_ `replace` statements (_see above_) to become this:
```js
function insert_person (data, callback) {
  connect( function insert_person_after_connected () {
    const { name, username, bio, worksfor, location, website, uid,
      stars, followers, following, contribs } = data; // destructured assignment
    const recent_activity = utils.recent_activity(data); // calculate recent activity for user
    const query = `INSERT INTO people (name, username, bio, worksfor, location,
      website, uid, stars, followers, following, contribs, recent_activity)
      VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12)`;
    const values = [name, username, bio, worksfor, location, website, uid,
      stars, followers, following, contribs, recent_activity];

    PG_CLIENT.query(query, values, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}
```
This is not "_perfect_" by _any_ means but it's a _lot_ better!

# Todo:
+ [x] update all the `insert` functions in the example #51 to use `pg` built-in parameterisation.

**Comment by nelsonic** [comment]
our `insert_org` function was:
```js
function insert_org (data, callback) {
  const fields = '(' + [ "url", "name", "description", "location",
  "website", "email", "pcount", "uid"].join(',') + ')';
  connect( function insert_org_after_connected () {
    const { url,name,description,location,website,email,pcount,uid } = data;
    // console.log(fields, name, username, company, uid, location);
    const placeholders = '%L, %L, %L, %L, %L, %L, $p, $1'
    let q = escape('INSERT INTO orgs %s VALUES (' + placeholders + ')',
      fields, url,name,description,location,website,email);
      // for some reason pg-escape does not play well with integers ...
      // see: https://github.com/segmentio/pg-escape/issues/15
      // so we are manually replacing the values:
      q = q.replace('$1', parseInt(uid, 10));
      q = q.replace('$p', parseInt(pcount, 10));
    // console.log('L93: query:', q);

    PG_CLIENT.query(q, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}
```
Simplified to:

```js
function insert_org (data, callback) {
  connect( function insert_org_after_connected () {
    const { url,name,description,location,website,email,pcount,uid } = data;
    const query = `INSERT INTO orgs
    (url, name, description, location, website, email, pcount, uid)
    VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`;
    const values = [url,name,description,location,website,email,pcount,uid];

    PG_CLIENT.query(query, values, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}
```
Cleaner and less "_explaining_" in the comments. 

I'm not _sure_ about creating the _local_ variables based on the values in the `data`Object ... 🤔 
Is this better: 
```js
function insert_org (data, callback) {
  connect( function insert_org_after_connected () {
    const query = `INSERT INTO orgs
    (url, name, description, location, website, email, pcount, uid)
    VALUES ($1, $2, $3, $4, $5, $6, $7, $8)`;
    const values = [data.url, data.name, data.description, data.location, 
      data.website, data.email, data.pcount, data.uid];

    PG_CLIENT.query(query, values, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}

```
What do _you_ think...?

**Comment by nelsonic** [comment]
our `insert_repo` function _was_:
```js
function insert_repo (data, callback) {

  const fields = '(' + [ "url", "description", "website", "tags", "langs",
  "watchers", "stars", "forks", "commits"].join(',') + ')';
  // console.log('db.js:L105: insert_repo > data', Object.keys(data).join(','));
  connect( function insert_repo_after_connected () {
    const { url, description, website, tags, langs, // string data
      watchers, stars, forks, commits} = data; // int data
    // console.log(fields, name, username, company, uid, location);
    const placeholders = '%L, %L, %L, %L, %L, $w, $s, $f, $c'
    let q = escape('INSERT INTO repos %s VALUES (' + placeholders + ')',
      fields, url, description, website, tags, langs.join(','));
      // for some reason pg-escape does not play well with integers ...
      // see: https://github.com/segmentio/pg-escape/issues/15
      // so we are manually replacing the values:
      q = q.replace('$w', parseInt(watchers, 10))
        .replace('$s', parseInt(stars, 10))
        .replace('$f', parseInt(forks, 10))
        .replace('$c', parseInt(commits, 10));

    // console.log('L121: query:', q);

    PG_CLIENT.query(q, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}
```
And is now _much_ cleaner:
```js
function insert_repo (data, callback) {
  connect( function insert_repo_after_connected () {
    const { url, description, website, tags, langs,
      watchers, stars, forks, commits} = data;
    const query = `INSERT INTO repos
    (url, description, website, tags, langs, watchers, stars, forks, commits)
    VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)`;
    const values = [url, description, website, tags, langs,
      watchers, stars, forks, commits];

    PG_CLIENT.query(query, values, function(error, result) {
      utils.log_error(error, data, new Error().stack);
      return insert_next_page (data, callback);
    });
  });
}
```
✅

**Comment by nelsonic** [comment]
Included in PR #51
